src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -586,6 +588,123 @@ def query_for_quota_alert(self):
+                devid = f.readline()
+            return devid[:-1]
+
+        def get_ro(zvol_dev):
+            with open(f'/sys/block/{zvol_dev}/ro', 'r') as f:
+                ro = f.readline()
+            return ro[:-1] == '1'
+
+        def get_attachment(zvol_vdev, data):
+            out = None
+            for method, attachment in data.items():
+                val = attachment.pop(zvol_vdev, None)
+                if val is not None:
+                    out = {
+                        'method': method,
+                        'data': val
+                    }
+                    break
+
+            return out
+
+        def get_zvols(info_level):
+            out = {}
+            zvol_path = '/dev/zvol/'
+            iscsi_offset = len('zvol/')
+            do_get_size = 'SIZE' in info_level
+            do_get_dev = 'DEVID' in info_level
+            do_get_ro = 'RO' in info_level
+            do_get_attachment = 'ATTACHMENT' in info_level
+
+            if do_get_attachment:
+                """
+                Normalize all places where we store zvol info so that
+                indexed by dataset name for quick lookup
+                """
+                extents = self.middleware.call_sync('iscsi.extent.query', [('type', '=', 'DISK')])
+                iscsi_zvols = {
+                    i['path'][iscsi_offset:]: i for i in extents
+                }
+
+                vm_devices = self.middleware.call_sync('vm.device.query', [['dtype', '=', 'DISK']])
+                vm_zvols = {
+                    zvol_path_to_name(i['attributes']['path']): i for i in vm_devices
+                }
+                attachment_data = {
+                    'iscsi.extent.query': iscsi_zvols,
+                    'vm.devices.query': vm_zvols
+                }
+
+            for root, dirs, files in os.walk(zvol_path):
+                if not files:
+                    continue
+
+                for file in files:
+                    path = root + '/' + file
+                    zvol_name = zvol_path_to_name(path)
+                    dev_name = os.readlink(path).split('/')[-1]
+
+                    out.update({
+                        zvol_name: {
+                            'path': path,
+                            'name': zvol_name,
+                            'dev': dev_name,
+                        }
+                    })
+
+                    if do_get_size is True:
+                        out[zvol_name]['size'] = get_size(dev_name)
+
+                    if do_get_dev is True:
+                        out[zvol_name]['devid'] = get_devid(dev_name)
+
+                    if do_get_ro is True:
+                        out[zvol_name]['ro'] = get_ro(dev_name)
+
+                    if do_get_attachment:
+                        out[zvol_name]['attachment'] = get_attachment(zvol_name, attachment_data)
+
+            return out
+
+        opts = options or {}
+        info_level = opts.get('additional_information', [])
```

Comment:
Having an `@accepts` decorator with `Dict` schema will reduce the number of these `or {}` and `.get(..., [])` and prevent typos in key names or enum values

Reply:
Switched to list. Put another arg in accepts decorator for the filterable version.